# March17FeaturedEngineering1Assignement

## Q1: What are missing values in a dataset? Why is it essential to handle missing values? Name some algorithms that are not affected by missing values.

### Missing values are values in a data set which are not present and are blank or null. It is essential to handle missing values. Missing data are a common occurrence and can have a significant effect on the conclusions that can be drawn from the data.

### Naive Bayes
## k-NN

## Q2: List down techniques used to handle missing data. Give an example of each with python code.

Techniques Used to handle missing data are-
1) Dropping or removing missing values
2) Imputing the missing values


### 1) Dropping missing values from data set
### We can remove missing values from a data set to handle them like


In [2]:
import seaborn as sns

In [3]:
titanicData = sns.load_dataset("titanic")

In [4]:
titanicData

survived  pclass     sex   age  sibsp  parch     fare embarked   class  \
0           0       3    male  22.0      1      0   7.2500        S   Third   
1           1       1  female  38.0      1      0  71.2833        C   First   
2           1       3  female  26.0      0      0   7.9250        S   Third   
3           1       1  female  35.0      1      0  53.1000        S   First   
4           0       3    male  35.0      0      0   8.0500        S   Third   
..        ...     ...     ...   ...    ...    ...      ...      ...     ...   
886         0       2    male  27.0      0      0  13.0000        S  Second   
887         1       1  female  19.0      0      0  30.0000        S   First   
888         0       3  female   NaN      1      2  23.4500        S   Third   
889         1       1    male  26.0      0      0  30.0000        C   First   
890         0       3    male  32.0      0      0   7.7500        Q   Third   

       who  adult_male deck  embark_town alive  alone  
0      man        True  NaN  Southampton    no  False  
1    woman       False    C    Cherbourg   yes  False  
2    woman       False  NaN  Southampton   yes   True  
3    woman       False    C  Southampton   yes  False  
4      man        True  NaN  Southampton    no   True  
..     ...         ...  ...          ...   ...    ...  
886    man        True  NaN  Southampton    no   True  
887  woman       False    B  Southampton   yes   True  
888  woman       False  NaN  Southampton    no  False  
889    man        True    C    Cherbourg   yes   True  
890    man        True  NaN   Queenstown    no   True  

[891 rows x 15 columns]

In [5]:
import pandas as pd

In [6]:
df = pd.DataFrame(titanicData)

In [8]:
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [18]:
df.isnull().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

### Here we can see that deck has 688 missing values and age has 177 missing values

In [19]:
## Delete the rows or data point to handle missing values

In [20]:
df.shape

(891, 15)

In [21]:
droppedDF = df.dropna()

In [22]:
droppedDF.shape

(182, 15)

In [23]:
#We can delete column wise

In [24]:
droppedCw = df.dropna(axis=1)

In [25]:
droppedCw.shape

(891, 11)

### 2) Handling missing values by imputing values

i) Imputing mean values\
We can replace missing values with mean

In [34]:
df['Age_mean'] =df['age'].fillna(df['age'].mean())

In [35]:
df[['age','Age_mean']]

age   Age_mean
0    22.0  22.000000
1    38.0  38.000000
2    26.0  26.000000
3    35.0  35.000000
4    35.0  35.000000
..    ...        ...
886  27.0  27.000000
887  19.0  19.000000
888   NaN  29.699118
889  26.0  26.000000
890  32.0  32.000000

[891 rows x 2 columns]

ii) imputing with median

In [37]:
df['Age_median'] =df['age'].fillna(df['age'].median())

In [38]:
df[['age','Age_mean','Age_median']]

age   Age_mean  Age_median
0    22.0  22.000000        22.0
1    38.0  38.000000        38.0
2    26.0  26.000000        26.0
3    35.0  35.000000        35.0
4    35.0  35.000000        35.0
..    ...        ...         ...
886  27.0  27.000000        27.0
887  19.0  19.000000        19.0
888   NaN  29.699118        28.0
889  26.0  26.000000        26.0
890  32.0  32.000000        32.0

[891 rows x 3 columns]

 iii) Mode Imputation Technqiue--Categorical values

In [40]:
df['embarked'].isna().sum()

2

In [41]:
df['embarked'].unique()

array(['S', 'C', 'Q', nan], dtype=object)

In [47]:
df['embarked'].mode()[0]

'S'

In [48]:
df['mode_embarked'] = df['embarked'].fillna(df['embarked'].mode()[0])

df[df['embarked'].isna()]

## Q3: Explain the imbalanced data. What will happen if imbalanced data is not handled?

### Imbalanced data refers to those types of datasets where the target class has an uneven distribution of observations, i.e one class label has a very high number of observations and the other has a very low number of observations

### If we do not handle imbalance data the result will be baised

## Q4: What are Up-sampling and Down-sampling? Explain with an example when up-sampling and down-sampling are required.

### Upsmpling is the technique to add more data points to the datasets having minority class and DOwn-sampling is the technique to reduce the number of data points of mjority class to balance the imbalance data

### When performance is required we can use down sampling and when accuracy is concerned we should use upsampling

## Q5: What is data Augmentation? Explain SMOTE.

### Data augmentation is a technique of artificially increasing the training set by creating modified copies of a dataset using existing data. It includes making minor changes to the dataset or using deep learning to generate new data points.

### SMOTE (Synthetic Minority Over-sampling Technique) is a technique used in machine learning to address imbalanced datasets where the minority class has significantly fewer instances than the majority class. SMOTE involves generating synthetic instances of the minority class by interpolating between existing instances.

## Q6: What are outliers in a dataset? Why is it essential to handle outliers?

### Outliers are the data point in the data set which lies out side lower fence and upper fence. It is essential to handle outliers  because they hampers the results and result will not be accurate and data will be imbalance.

## Q7: You are working on a project that requires analyzing customer data. However, you notice that some of the data is missing. What are some techniques you can use to handle the missing data in your analysis?

Techniques Used to handle missing data are-\
1) Dropping or removing missing values row wise or column wise
2) Imputing the missing values with mean median or mode

## Q8: You are working with a large dataset and find that a small percentage of the data is missing. What are some strategies you can use to determine if the missing data is missing at random or if there is a pattern to the missing data?

### If missing data is dependent on observed data then it is Missing at Random. If missing values depends on the value of the missing data itself then there is a pattern to the missing data.

## Q9: Suppose you are working on a medical diagnosis project and find that the majority of patients in the dataset do not have the condition of interest, while a small percentage do. What are some strategies you can use to evaluate the performance of your machine learning model on this imbalanced dataset?

### We can use upsampling or down sampling strategies to balance this data set.
### We can upsample minority data set or down sample by randomly selected majority data set

## Q10: When attempting to estimate customer satisfaction for a project, you discover that the dataset is unbalanced, with the bulk of customers reporting being satisfied. What methods can you employ to balance the dataset and down-sample the majority class?

### Sklearn.utils.resample can be used for both down samplings the majority class and upsample minority class instances.

## Q11: You discover that the dataset is unbalanced with a low percentage of occurrences while working on a project that requires you to estimate the occurrence of a rare event. What methods can you employ to balance the dataset and up-sample the minority class?

### SMOTE (Synthetic Minority Over-sampling Technique) is a technique used in machine learning to address imbalanced datasets where the minority class has significantly fewer instances than the majority class. SMOTE involves generating synthetic instances of the minority class by interpolating between existing instances.